File: 46.全排列.go
/*
 * @lc app=leetcode.cn id=46 lang=golang
 *
 * [46] 全排列
 */

// @lc code=start
func permute(nums []int) (ret [][]int) {
	var backtrack func([]int)
	backtrack = func(path []int) {
		if len(path) == len(nums) {
			ret = append(ret, path)
			return
		}

		for i := 0; i < len(nums); i++ {
			if contains(path, nums[i]) {
				continue
			}
			backtrack(append(path, nums[i]))
		}
	}
	backtrack([]int{})
	return
}
func contains(arr []int, target int) bool {
	for _, v := range arr {
		if v == target {
			return true
		}
	}
	return false
}
// @lc code=end

File: 46.全排列.rs
/*
 * @lc app=leetcode.cn id=46 lang=rust
 *
 * [46] 全排列
 */

// @lc code=start
impl Solution {
    pub fn permute(nums: Vec<i32>) -> Vec<Vec<i32>> {
        let mut ret: Vec<Vec<i32>> = Vec::new();

        fn backtrack(path: &mut Vec<i32>, nums: &Vec<i32>, ret: &mut Vec<Vec<i32>>) {
            if path.len() == nums.len() {
                ret.push(path.clone());
                return;
            }

            for i in 0..nums.len() {
                if path.contains(&nums[i]) {
                    continue;
                }
                path.push(nums[i]);
                backtrack(path, nums, ret);
                path.pop();
            }
        }

        backtrack(Vec::new().as_mut(), &nums, &mut ret);
        ret
    }
}
// @lc code=end

